### PR TITLE
Add PerceiverLayer implementation

### PR DESCRIPTION
**Pull request by tshimko-nv**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Add Perceiver Layer implementation to NeMo LLM collection for efficient processing of long sequences using latent attention.

**Collection**: [Note which collection this PR will affect]

LLM

# Changelog

- Added new PerceiverConfig class extending TransformerConfig with Perceiver-specific parameters
- Implemented PerceiverLayer class that processes input sequences through:
  - Cross-attention between latent array and input sequence
  - Self-attention on latent array
  - MLP processing on latent states
- Added PerceiverLayerSubmodules dataclass for configurable module specifications
- Provided example perceiver_layer_spec ModuleSpec for standard Perceiver configuration
- Integrated with Megatron-Core transformer components and Transformer Engine optimizations

# Usage

```python
# Configure and instantiate a Perceiver Layer
config = PerceiverConfig(
    num_latents=256,  # Number of latent vectors
    hidden_size=1024,
    num_attention_heads=16
)

# Create Perceiver layer using the provided spec
perceiver = PerceiverLayer(
    config=config,
    submodules=perceiver_layer_spec.submodules
)

# Forward pass
output = perceiver(
    latent_states=latents,      # shape: [num_latents, batch_size, hidden_size]
    input_sequence=inputs,      # shape: [seq_length, batch_size, hidden_size]
    cross_attention_masks=cross_masks,
    self_attention_mask=self_mask
)
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/perceiver/model/perceiver_layer.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+            requires_grad=latent_states.requires_grad,
+            keep_graph=True
+        )
+        print(f"Final output: {output.shape}")
+
+        return output
+
+
+# Example ModuleSpec for Perceiver layer
+perceiver_layer_spec = ModuleSpec(
```
Comment (marcromeyn) [inline nemo/collections/llm/perceiver/model/perceiver_layer.py:201]: Do you think it would be useful to have a non-TE option as well? If so, it would be good to guard the TE imports.

File: nemo/collections/llm/perceiver/model/perceiver_layer.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+                cross-attention. Each mask has shape [batch_size, 1, 1, seq_len].
+            self_attention_mask (torch.Tensor): Attention mask for self-attention on latents,
+                with shape [batch_size, 1, 1, latent_len].
+            inference_params (Optional[dict]): Optional parameters to optimize inference-time computation.
+
+        Returns:
+            torch.Tensor: Updated latent states after processing through the layer,
+                with shape [latent_len, batch_size, hidden_size].
+        """
+        print("\nShape tracking through forward pass:")
```
Comment (marcromeyn) [inline nemo/collections/llm/perceiver/model/perceiver_layer.py:163]: I would change all print statements to logging.info

File: tests/collections/llm/perceiver/test_perceiver_layer.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
+import pytest
```
Comment (marcromeyn) [inline tests/collections/llm/perceiver/test_perceiver_layer.py:1]: Copyright is missing

File: tests/collections/llm/perceiver/test_perceiver_layer.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+import torch
+from typing import Optional
```
[the remark this answers is not in the document]
Comment (tshimko-nv) [reply tests/collections/llm/perceiver/test_perceiver_layer.py]: Done.

File: tests/collections/llm/perceiver/test_perceiver_layer.py (hunk 3 above)
Comment (tshimko-nv) [reply tests/collections/llm/perceiver/test_perceiver_layer.py:15]: Added, thanks for catching this!

File: nemo/collections/llm/perceiver/model/perceiver_layer.py (hunk 2 above)
Comment (tshimko-nv) [reply nemo/collections/llm/perceiver/model/perceiver_layer.py]: I removed these, they were left over from dev testing.

File: nemo/collections/llm/perceiver/model/perceiver_layer.py (hunk 1 above)
Comment (pstjohn) [reply nemo/collections/llm/perceiver/model/perceiver_layer.py:182]: Any reason? We haven't been diligent to import-guard TE in the rest of the framework; our other models (ESM-2, Geneformer, etc.) have it as a hard dependency)
